File: view/src/lib/types/Spell.ts
import { Class, MagicSchool, SpellLevel } from "$types";

export type Spell = {
  id?: string;
  name?: string;
  school?: MagicSchool;
  level?: SpellLevel;
  casting_time?: string;
  verbal: boolean;
  somatic: boolean;
  material?: string;
  material_consumed: boolean;
  ritual: boolean;
  concentration: boolean;
  duration?: string;
  range?: string;
  area?: string;
  shape?: string;
  classes: Class[];
  description?: string;
  at_higher_levels?: string;
};

export const SpellActions = {
  EmptySpell: (): Spell => ({
    id: undefined,
    name: undefined,
    school: undefined,
    level: undefined,
    casting_time: undefined,
    verbal: false,
    somatic: false,
    material: undefined,
    material_consumed: false,
    ritual: false,
    concentration: false,
    duration: undefined,
    range: undefined,
    area: undefined,
    shape: undefined,
    classes: [],
    description: undefined,
    at_higher_levels: undefined,
  }),
};

File: api/src/types/spell.rs
use super::{Class, MagicSchool, SpellLevel};
use uuid::Uuid;

#[derive(Debug, serde::Serialize, serde::Deserialize, sqlx::FromRow)]
pub struct Spell {
    pub id: Option<Uuid>,
    pub name: String,
    pub school: MagicSchool,
    pub level: SpellLevel,
    pub verbal: bool,
    pub somatic: bool,
    pub material: Option<String>,
    pub material_consumed: bool,
    pub ritual: bool,
    pub concentration: bool,
    pub casting_time: String,
    pub duration: String,
    pub range: String,
    pub area: String,
    pub shape: Option<String>,
    pub description: String,
    pub at_higher_levels: Option<String>,
    pub classes: Vec<Class>,
}
